### PR TITLE
docs(WebWorkers): Add overview of WebWorker design

### PR DESCRIPTION
**Pull request by jteplitz**

### CODE REVIEW

File: modules/angular2/docs/web_workers/web_workers.md [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+infrastructure that angular uses to communicate between the main process and the
+worker. This infrastructure can be modified by an application developer to
+enable driving an angular 2 application from an iFrame, different window / tab,
+server, etc..
+
+## Introduction
+WebWorker support in Angular2 is designed to make it easy to leverage parallelization in your web application.
+When you choose to run your application in a WebWorker angular runs both your application's logic and the 
+majority of the core angular framework in a WebWorker. 
+By offloading as much code as possible to the
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: You are a master of suspense ;)

File: modules/angular2/docs/web_workers/web_workers.md [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+import "package:angular2/web_worker/worker.dart";
+import "package:angular2/src/core/reflection/reflection.dart";
+import "package:angular2/src/core/reflection/reflection_capabilities.dart";
+
+@Component({
+  selector: "hello-world"
+})
+@View({
+  template: "<h1>Hello {{name}}</h1>"
+})
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: The annotations above use TS syntax.

File: modules/angular2/docs/web_workers/web_workers.md [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+      console.log(message); // will print "hello from the UI"
+    });
+  }
+}
+```
+
+This example is nearly identical in Dart, and is included below for reference:
+```Dart
+// index.dart, which is running on the UI.
+import 'package:angular2/web_workers/worker';
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: `worker.dart`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+  var instance = bootstrap("background_index.dart");
+  var bus = instance.bus;
+  bus.initChannel("My Custom Channel");
+  bus.to("My Custom Channel").add("hello from the UI");
+}
+
+```
+```Dart
+// background_index.dart, which is running on the WebWorker
+import 'package:angular2/web_worker/worker';
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: `worker.dart`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+@Component({...})
+@View({...})
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: Use Dart annotation syntax

File: modules/angular2/docs/web_workers/web_workers.md [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+class MyComponent {
+  MyComponent (MessageBus bus) {
+    bus.initChannel("My Custom Channel");
+    bus.from("My Custom Channel").listen((message) => {
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: Remove `=>`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+    broker.runOnService(methodInfo, PRIMTIVE).then((result: string) => {
+      // result will be equal to the return value of doCoolThing(value) that ran on the UI.
+    });
+  }
+}
+```
+#### Using the MessageBroker in Dart
+```Dart
+// index.dart, which is running on the UI with a method that we want to expose to a WebWorker
+import 'package:angular2/web_worker/ui';
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: `ui.dart`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+  var broker = instance.app.createServiceMessageBroker("My Broker Channel");
+  
+  // assume we have some function doCoolThings that takes a String argument and returns a Future<String>
+  broker.registerMethod("awesomeMethod", [PRIMITIVE], (String arg1) => doCoolThing(arg1), PRIMITIVE);
+}
+
+```
+```Dart
+// background.dart, which is running on a WebWorker and wants to execute a method on the UI
+import 'package:angular2/web_worker/worker';
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: `worker.dart`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+
+@Component(...)
+@View(...)
+export class MyComponent {
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: Remove `export`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+  constructor(ClientMessageBrokerFactory brokerFactory) {
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: `/constructor/MyComponent/`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+    var broker = brokerFactory.createMessageBroker("My Broker Channel");
+    
+    var arguments = [new FnArg(value, PRIMITIVE)];
+    var methodInfo = new UiArguments("awesomeMethod", arguments);
+    broker.runOnService(methodInfo, PRIMTIVE).then((String result) => {
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: remove `=>`

File: modules/angular2/docs/web_workers/web_workers.md [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+In JavaScript / TypeScript this means they must be serializable via JavaScript's 
+[structured cloning algorithim](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm).
+
+In Dart this means they must be valid messages that can be passed through a
+[SendPort](https://api.dartlang.org/1.12.1/dart-isolate/SendPort/send.html).
+
+
+### MessageBus and Zones
+The MessageBus API includes support for [zones](http://www.github.com/angular/zone.js). 
+A MessageBus can be attached to a specific zone  (by calling `attachToZone). Then specific channels can be
```
Comment (yjbanov) [inline modules/angular2/docs/web_workers/web_workers.md]: forgot closing backtick after `attachToZone`

**Comment by yjbanov** [comment]
/cc @naomiblack @kwalrath 

This is a very user-friendly design doc. We should turn it into a page on angular.io.

**Comment by webteckie** [comment]
fyi...there's an incomplete sentence in the introduction:  "By offloading as much code as possible to the"

**Comment by jteplitz602** [comment]
Fixed. Thanks Carlos!

File: modules/angular2/docs/web_workers/web_workers.md (hunk 1 above)
Comment (gkalpak) [inline modules/angular2/docs/web_workers/web_workers.md]: Epic comment (lol)

File: modules/angular2/docs/web_workers/web_workers.md [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+  <body>
+    <hello-world></hello-world>
+    <script>System.import("index")</script>
+  </body>
+</html>
+```
+```TypeScript
+// index.js
+import {bootstrap} from "angular2/web_worker/ui";
+bootstrap(loader.js);
```
Comment (gkalpak) [inline modules/angular2/docs/web_workers/web_workers.md]: Should this be `'loader.js'` ?

File: modules/angular2/docs/web_workers/web_workers.md [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+bus.initChannel("My Custom Channel");
+```
+```TypeScript
+// background_index.ts, which is running on the WebWorker
+import {MessageBus} from 'angular2/web_worker/worker';
+@Component({...})
+@View({...})
+export class MyComponent {
+  constructor (bus: MessageBus) {
+    bus.initChannel("My CustomChannel");
```
Comment (gkalpak) [inline modules/angular2/docs/web_workers/web_workers.md]: _CustomChannel_  -->  _Custom Channel_ ?

File: modules/angular2/docs/web_workers/web_workers.md [hunk 15]
```diff
@@ -0,0 +1,476 @@
...
+The MessageBroker is a higher level messaging abstraction that sits on top of the MessageBus. It is used when you
+want to execute code on the other side of a runtime boundary and may want to receive the result.
+There are two types of MessageBrokers. The `ServiceMessageBroker` is used by the side that actually performs
+an operation and may return a result. Conversely, the `ClientMessageBroker` is used by the side that requests that
+an operation be performed and may want to receive the result.
+
+### Using the MessageBroker In Your Application
+To use MessageBrokers in your application you must initialize both a `ClientMessageBroker` and a
+`ServiceMessageBroker` on the same channel. You can then register methods that the `ServiceMessageBroker` should
+listen the instruct the `ClientMessageBroker` to run those methods. Below is a lightweight example of using
```
Comment (gkalpak) [inline modules/angular2/docs/web_workers/web_workers.md]: _listen the instruct the `ClientMessageBroker`_  -->  ??

Comment (jteplitz602) [inline modules/angular2/docs/web_workers/web_workers.md]: Fixed this and other typos in 6a4a8a60abd4d5eb5dd93e3c9787e2b8cd5f48c4

Thanks for pointing them out Georgios!

**Pull request closed without merging**